src/choker.cpp

```diff
@@ -317,28 +313,28 @@ namespace {
 			// it purely based on the current state of our peers.
 			upload_slots = 0;
 
-			// TODO: use an incremental partial_sort() here. We don't need
-			// to sort the entire list
+			int rate_threshold = sett.get_int(settings_pack::rate_choker_initial_threshold);
 
-			// TODO: make the comparison function a free function and move it
-			// into this cpp file
-			std::sort(peers.begin(), peers.end()
-				, std::bind(&upload_rate_compare, _1, _2));
+			auto begin = peers.begin();
 
-			// TODO: make configurable
-			int rate_threshold = 1024;
-
-			for (auto const p : peers)
+			for (int i = 0; i < int(peers.size()); ++i, ++begin)
 			{
+				// just make sure the next element is sorted
+				std::nth_element(begin, begin, peers.end()
+					, std::bind(&upload_rate_compare, _1, _2));
+
+				auto const* p = *begin;
+
 				int const rate = int(p->uploaded_in_last_round()
 					* 1000 / total_milliseconds(unchoke_interval));
 
-				if (rate < rate_threshold) break;
+				// always have at least 1 unchoke slot
+				if (rate < rate_threshold && i >= 1) break;
 
 				++upload_slots;
 
-				// TODO: make configurable
-				rate_threshold += 1024;
+				// double the threshold
+				rate_threshold += rate_threshold;
```

Comment:
Maybe check 
if ( rate_threshold <= 1048576)
rate_threshold += rate_threshold;
to prevent the rate being increased infinitely.

Reply:
could you provide some more rationale for that? That specific logic you propose would have a pretty weird singularity right at that point..

Reply:
Let’s say it’d take 8 MiB/s to unchoke the next peer. But the next peer can only reach 7.9 MiB/s because of the bottleneck on my end. Then I’ll be wasting that 7.9MiB/s upload capacity. Whereas if we had a max rate limit of lets say 1 MiB/s we would lose max 0.99MiB/s upload capacity.

Reply:
but you have to weigh that against the negative effects of having too many unchoke slots too. If all we care about is using all our upload capacity, unlimited unchoke slots would waste even less. but if your storage is spinning rust, of course, you may have worse performance with too many unchoke slots. But you may also drop below other peer's reciprocation rate.

I'm not saying the current algorithm is perfect, but I think more analysis is required.

For example, perhaps instead of doubling the threshold, it could increase by 50%.

Reply:
in fact, this might be an interesting telemetry metric for qbt. What is the typical reciprocation rate? But I also think it's important to have an algorithm that's robust to faster internet connections

Reply:
Maybe after reaching a certain rate you can change the multiplier to be 0.5x instead of 2x.